website/server/libs/googlePayments.js

```diff
@@ -0,0 +1,156 @@
+import shared from '../../common';
+import iap from './inAppPurchases';
+import payments from './payments';
+import {
+  NotAuthorized,
+  BadRequest,
+} from './errors';
+import { model as IapPurchaseReceipt } from '../models/iapPurchaseReceipt';
+import {model as User } from '../models/user';
+import moment from 'moment';
+
+let api = {};
+
+// TODO missing tests
+
+
+api.verifyGemPurchase = async function verifyGemPurchase (user, receipt, signature, headers) {
+  await iap.setup();
+
+  let testObj = {
+    data: receipt,
+    signature,
+  };
+
+  let googleRes = await iap.validate(iap.GOOGLE, testObj);
+
+  let isValidated = iap.isValidated(googleRes);
+  if (!isValidated) throw new NotAuthorized('INVALID_RECEIPT');
+
+  let receiptObj = JSON.parse(testObj.data); // passed as a string
+  let token = receiptObj.token || receiptObj.purchaseToken;
+
+  let existingReceipt = await IapPurchaseReceipt.findOne({
+    _id: token,
+  }).exec();
+  if (existingReceipt) throw new NotAuthorized('RECEIPT_ALREADY_USED');
+
+  await IapPurchaseReceipt.create({
+    _id: token,
+    consumed: true,
+    userId: user._id,
+  });
+
+  let amount;
+
+  switch (receiptObj.productId) {
+    case 'com.habitrpg.android.habitica.iap.4gems':
+      amount = 1;
+      break;
+    case 'com.habitrpg.android.habitica.iap.20.gems':
```

Comment:
yes. that's an old format we used. it shouldn't be in usage anymore, but it's there for legacy purposes in case someone still has a receipt with that